finality-aleph/src/validator_network/outgoing.rs

```diff
@@ -10,39 +10,31 @@ use crate::{
     validator_network::{
         protocol_negotiation::{protocol, ProtocolNegotiationError},
         protocols::ProtocolError,
-        Data, Dialer,
+        ConnectionInfo, Data, Dialer, PeerAddressInfo,
     },
 };
 
 enum OutgoingError<A: Data, ND: Dialer<A>> {
     Dial(ND::Error),
-    ProtocolNegotiation(ProtocolNegotiationError),
-    Protocol(ProtocolError),
+    ProtocolNegotiation(PeerAddressInfo, ProtocolNegotiationError),
+    Protocol(PeerAddressInfo, ProtocolError),
 }
 
 impl<A: Data, ND: Dialer<A>> Display for OutgoingError<A, ND> {
     fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
         use OutgoingError::*;
         match self {
             Dial(e) => write!(f, "dial error: {}", e),
-            ProtocolNegotiation(e) => write!(f, "protocol negotiation error: {}", e),
-            Protocol(e) => write!(f, "protocol error: {}", e),
+            ProtocolNegotiation(addr, e) => write!(
+                f,
+                "protocol negotiation error: {}, peer address info: {}",
```

Comment:
Similarly I would go with something like "protocol negotiation with {} error: {}" and for consistency also below.

Reply:
changed to "Outgoing connection to {} failed, protocol negotiation error: {}.", and "Outgoing connection to {} failed, protocol error: {}.", since I was unable to make the second sentence look good in the proposed style

Reply:
Unfortunately this is not supposed to be a whole message, just the part that corresponds to this specific error – in particular whatever you write here will end up formatted in `info!(target: "validator-network", "Outgoing connection to {} failed: {}, will retry after {}s.", peer_id, e, RETRY_DELAY.as_secs());`, so you would get quite some repetition. I think this will mostly be fine if you remove "Outgoing" from these strings?

Reply:
but then it would become "Outgoing connection to {} failed: connection to {} failed, protocol negotiation error: {}, will retry after {}s.", which doesn't sound good...

I'd say in that case "peer address info: {}, protocol negotiation error: {}" is much better, what do you think?

Reply:
If doesn't really flow great, the thing you wrote out doesn't sound that bad once you notice the two `{}` there are different things – the first one is a peer id, the second is an address.

How about "communication with {} failed, protocol[ negotiation] error: {}"?

Reply:
fixed